File: lintcode/easy/space_replacement/py/space_replacement.py
# coding:utf-8
'''
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/space-replacement
@Language: Python
@Datetime: 17-02-15 13:39
'''

class Solution:
    # @param {char[]} chars: An array of Char
    # @param {int} length: The true length of the string
    # @return {int} The true length of new string
    def replaceBlank(self, chars, length):
        if chars == None:
            return None
        
        buff = chars[:length]
        j    = 0
        
        for i in range(len(buff)):
            if buff[i] == ' ':
                for c in '%20':
                    chars[j] = c
                    j += 1
            else:
                chars[j] = buff[i]
                j += 1
                
        return j

File: lintcode/easy/space_replacement/cpp/space_replacement.cpp
/*
@Copyright:LintCode
@Author:   lilsweetcaligula
@Problem:  http://www.lintcode.com/problem/space-replacement
@Language: C++
@Datetime: 17-02-15 15:49
*/

class Solution {
public:
    /**
     * @param string: An array of Char
     * @param length: The true length of the string
     * @return: The true length of new string
     */
    int replaceBlank(char string[], int length) 
    {
        static const std::string repl { "%20" };
        const std::vector<char> buffer { &string[0], &string[length] };
        int index { 0 };
        
        for (const auto ch : buffer) {
            if (ch == ' ') {
                for (const auto replCh : repl) {
                    string[index++] = replCh;
                }
            } else {
                string[index++] = ch;   
            }
        }
        
        return index;
    }
};